### PR TITLE
[HMA] Move mock to API level

### PR DESCRIPTION
**Pull request by BarrettOlson**
Summary
---------

Connects more of the UI mocks to the python lambda API. The changes aren't easily apparent on the UI apart from where it nows says 'mocked' on the Dashboard and Signals Page. Most changes are likely to be refactored, consolidated, or extended once hooked up to real data.

Test Plan
---------

Tested new endpoints with UI and Postman. 
Screen Shot 2021-04-08 at 11 53 04 AM
Screen Shot 2021-04-08 at 11 52 58 AM

### CODE REVIEW

**Comment by Dcallies** [comment]
Please let me know when you want this reviewed, will ignore until it's up for review.

**Comment by Dcallies** [comment]
In general, would suggest breaking this up more - it's perfectly fine to have a single PR for just one API connected, even if it's just 10 or 30 lines.

**Comment by BarrettOlson** [comment]
> In general, would suggest breaking this up more - it's perfectly fine to have a single PR for just one API connected, even if it's just 10 or 30 lines.

yes, apologies. I can try to break this up if you want. the individual changes, because mocking, is involved didn't feel significant enough until it snowballed.

**Comment by Dcallies** [comment]
> 
> 
> > In general, would suggest breaking this up more - it's perfectly fine to have a single PR for just one API connected, even if it's just 10 or 30 lines.
> 
> yes, apologies. I can try to break this up if you want. the individual changes, because mocking, is involved didn't feel significant enough until it snowballed.

Yea, no worry at this point fine to leave together, mostly talking about followups. We talked about the cost of long iterations tending us towards bigger diffs as a reason to combine right now, but I'm worried it's a positive feedback cycle that leads to overall slower cycles.

File: hasher-matcher-actioner/hmalib/lambdas/api/api_root.py [hunk 1]
```diff
@@ -161,28 +202,47 @@ def gen_matches() -> t.List[MatchesResult]:
...
+    # TODO these mocked metadata should either be added to
+    # PDQMatchRecord or some other look up in the data model
+    MOCKED_METADATA: MatchDetailsMetaData = {
+        "type": "HASH_PDQ",
+        "tags": ["mocked_t1", "mocked_t2"],
+        "status": "MOCKED_STATUS",
+        "opinions": ["mocked_a1", "mocked_a2"],
+    }
+
+    MOCKED_ACTIONS = ["Mocked_False_Postive", "Mocked_Delete"]
```
Comment (Dcallies) [inline hasher-matcher-actioner/hmalib/lambdas/api/api_root.py:245]: .

File: hasher-matcher-actioner/hmalib/lambdas/api/api_root.py [hunk 2]
```diff
@@ -161,28 +202,47 @@ def gen_matches() -> t.List[MatchesResult]:
...
     if not content_id:
         return []
     table = dynamodb.Table(DYNAMODB_TABLE)
     records = PDQMatchRecord.get_from_content_id(
         table, f"{IMAGE_FOLDER_KEY}{content_id}"
     )
+    # TODO these mocked metadata should either be added to
+    # PDQMatchRecord or some other look up in the data model
+    MOCKED_METADATA: MatchDetailsMetaData = {
+        "type": "HASH_PDQ",
```
Comment (Dcallies) [inline hasher-matcher-actioner/hmalib/lambdas/api/api_root.py:239]: Nit: TypedDicts allow this:

```
  MatchDetailsMetaData(type="HASH_PDQ", tags=[...])
```

Which means you don't have to add a type for them.

File: hasher-matcher-actioner/hmalib/lambdas/api/api_root.py [hunk 3]
```diff
@@ -216,3 +278,93 @@ def gen_hash(content_id: str) -> t.Optional[HashResult]:
...
+def get_signals() -> t.List[SignalSourceSummary]:
+    """
+    TODO this should be updated to check ThreatExchangeConfig
+    based on what it finds in the config it should then do a s3 select on the files
+    https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.select_object_content
+    """
+    MOCKED_SIGNAL_TYPE1A: SignalSourceType = {"type": "HASH_PDQ", "count": 12456}
+    MOCKED_SIGNAL_TYPE1B: SignalSourceType = {"type": "MOCKED_TYPE1", "count": 456}
+    MOCKED_SIGNAL_TYPE2A: SignalSourceType = {"type": "HASH_PDQ", "count": 2456}
+    MOCKED_SIGNAL_TYPE2B: SignalSourceType = {"type": "MOCKED_TYPE2", "count": 956}
```
Comment (Dcallies) [inline hasher-matcher-actioner/hmalib/lambdas/api/api_root.py:303]: same note on pep8 stuff

File: hasher-matcher-actioner/hmalib/lambdas/api/api_root.py [hunk 4]
```diff
@@ -216,3 +278,93 @@ def gen_hash(content_id: str) -> t.Optional[HashResult]:
...
+class Dashboard(t.TypedDict):
+    hashes: DashboardCount
+    matches: DashboardCount
+    actions: DashboardCount
+    signals: DashboardCount
+    system_status: DashboardSystemStatus
+
+
+def get_dashboard_mock() -> Dashboard:
+    hashes: DashboardCount = {
```
Comment (Dcallies) [inline hasher-matcher-actioner/hmalib/lambdas/api/api_root.py:339]: same comment about the wrapper

File: hasher-matcher-actioner/hmalib/lambdas/api/api_root.py [hunk 5]
```diff
@@ -128,11 +128,52 @@ def hashes(key=None):
...
+    results = get_hash(key)
     logger.debug(results)
     return results if results else {}
 
 
+@app.get("/content_status/<key>")
+def content_status(key=None):
+    """
+    content status API endpoint:
+    currently returns mocked data
```
[the remark this answers is not in the document]
Comment (BarrettOlson) [reply hasher-matcher-actioner/hmalib/lambdas/api/api_root.py:140]: Exactly what this should return is a little dependent on the actioning framework and other tbd things. Just cutting the second line for now.

File: hasher-matcher-actioner/hmalib/lambdas/api/api_root.py [hunk 6]
```diff
@@ -161,28 +197,47 @@ def gen_matches() -> t.List[MatchesResult]:
...
             "signal_source": record.signal_source,
             "updated_at": record.updated_at.isoformat(),
-            "reactions": "TODO",
+            "reactions": "Mocked",
         }
         for record in records
     ]
 
 
+class MatchDetailsMetaData(t.TypedDict):
```
Comment (eriksendc) [inline hasher-matcher-actioner/hmalib/lambdas/api/api_root.py:206]: Nit... I think the official word is metadata, so `MatchDetailsMetadata` when using camel case.

File: hasher-matcher-actioner/hmalib/lambdas/api/api_root.py [hunk 7]
```diff
@@ -216,3 +273,93 @@ def gen_hash(content_id: str) -> t.Optional[HashResult]:
...
+    )
+    mocked_signal_source2 = SignalSourceSummary(
+        name="Mocked Signal Source 2",
+        signals=[mocked_signal_type2a, mocked_signal_type2b],
+        updated_at="mocked datetime2",
+    )
+    return [mocked_signal_source1, mocked_signal_source2]
+
+
+class DashboardCount(t.TypedDict):
```
Comment (jeberl) [inline hasher-matcher-actioner/hmalib/lambdas/api/api_root.py:313]: Why do we want all of these as TypedDicts and not just classes with the dict keys as class variables?

File: hasher-matcher-actioner/webapp/src/Api.jsx [hunk 8]
```diff
@@ -59,3 +71,9 @@ export async function uploadImage(file) {
     });
   };
 }
+
+export async function updateContentStatus(key, status) {
+  apiPost(`/content_status/${key}`, {
```
[the remark this answers is not in the document]
Comment (jeberl) [reply hasher-matcher-actioner/webapp/src/Api.jsx:76]: I would have done the exact same @BarrettOlson. FB born and bred

File: hasher-matcher-actioner/hmalib/lambdas/api/api_root.py (hunk 7 above)
Comment (BarrettOlson) [reply hasher-matcher-actioner/hmalib/lambdas/api/api_root.py:313]: I think that would be a fair amount more code and hard to get field level typing that way. TypedDict is the most minimal solutions I found in a quick search to have some sort of typing on our return data.

File: hasher-matcher-actioner/webapp/src/Signals.jsx [hunk 9]
```diff
@@ -1,80 +1,84 @@
...
+        </Collapse>
+      </div>
+    </>
+  );
+}
+
+function SignalSummary({summary}) {
+  return (
+    <>
+      <div className="col-md-12 mb-4">
```
[the remark this answers is not in the document]
Comment (BarrettOlson) [reply hasher-matcher-actioner/webapp/src/Signals.jsx]: updated all the files this PR touched to no longer use raw <div>

File: hasher-matcher-actioner/webapp/src/Dashboard.jsx [hunk 10]
```diff
@@ -1,117 +1,112 @@
...
 
-import React from 'react';
+import React, {useState, useEffect} from 'react';
 import {useHistory} from 'react-router-dom';
 
+import {fetchDashboardSummary} from './Api';
+
 export default function Dashboard() {
   const history = useHistory();
+  const [dashboardSummary, setDashboardSummary] = useState(null);
```
[the remark this answers is not in the document]
Comment (BarrettOlson) [reply hasher-matcher-actioner/webapp/src/Dashboard.jsx]: I wanted to keep it simple for the first pass but I found a way to just to the request in `<DashboardCard>` so I broke up the API as requested (don't love the resulting python from the change but we'll iterate)

**Review by eriksendc — APPROVED** [review]
LURV IT

**Pull request merged**